src/Instances/TH/Lift/Internal.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Instances.TH.Lift.Internal
+    ( liftText
+    , liftTextArray
+    , liftByteArray
+    , liftByteArrayPinned
+    , liftByteString
+    , liftPrimitiveVector
+    , liftStorableVector
+    ) where
+
+import           Control.Monad
+import           Control.Monad.Primitive (PrimMonad (..))
+import           Control.Monad.ST
+import qualified Data.ByteString as ByteString
+import qualified Data.ByteString.Unsafe as ByteString
+import           Data.Primitive.ByteArray
+import qualified Data.Text.Array as TArray
+import qualified Data.Text.Internal as Text
+import qualified Data.Vector.Primitive as Vector.Primitive
+import qualified Data.Vector.Storable as Vector.Storable
+import           Data.Word (Word8)
+import           Foreign.Storable
+import           Foreign.ForeignPtr
+import           GHC.Prim (Addr#, copyAddrToByteArray#, sizeofByteArray#)
+import           GHC.Ptr (Ptr(..), castPtr)
+import           GHC.Types (IO(..), Int(..))
+import           Language.Haskell.TH
+import           System.IO.Unsafe (unsafePerformIO)
+
+liftText :: Text.Text -> ExpQ
+liftText (Text.Text arr0 off len) =
```

Comment:
Oh I see, you want to only lift the part that is actually used, liftTextArray would lift the bytes before offset as well